### ISSUE TITLE
Unable to fetch Rust/Cargo dependencies through TLS-intercepting proxy

### ISSUE THREAD

**Issue by makefu** [NixOS/nixpkgs#89526]
**Describe the bug**
Nix is able to fetch source dependencies from github and other URLs but building a rust package fails when rust is trying to fetch the cargo dependencies.
`CURL_CA_BUNDLE` `GIT_SSL_CAINFO` and `SSL_CERT_FILE` contain paths to the certificate chain of the intercepting proxy.

**To Reproduce**
1. set up TLS-Intercepting proxy (or work in a company which uses this tech)
2. Set `CURL_CA_BUNDLE`, `GIT_SSL_CAINFO` and `SSL_CERT_FILE` to `/etc/ssl/certs/ca-bundle.crt`
3. `security.pki.certificates = [ (lib.readFile ./intercepting-proxy.pem ];`
4. build a rust package which is not cached by cache.nixos.org, e.g. this derivation: https://git.ingolf-wagner.de/palo/nixos-config/src/master/pkgs/navi/default.nix

**Expected behavior**
The package is built from source and can be used in configuration

**Screenshots**
```
building '/nix/store/dsi88ymq5wbm5ds86xxxnahkl1gk6p0f-navi-2.1.1-vendor.tar.gz.drv'...
unpacking sources
unpacking source archive /nix/store/4xhyh64vhlxxkfwpl7zbr1wcj0yxqafx-source
source root is source
patching sources
building
Updating crates.io index
warning: spurious network error (2 tries remaining): [56] Failure when receiving data from the peer; class=Net (12)
warning: spurious network error (1 tries remaining): [56] Failure when receiving data from the peer; class=Net (12)
error: failed to sync

Caused by:
failed to load pkg lockfile

Caused by:
failed to fetch https://github.com/rust-lang/crates.io-index

Caused by:
[56] Failure when receiving data from the peer; class=Net (12)
```

**Additional context**

**Notify maintainers**
@Ericson2314 @bhipple 

**Metadata**
- system: `"x86_64-linux"`
- host os: `Linux 5.4.35, NixOS, 20.09pre-git (Nightingale)`
- multi-user?: `no`
    - sandbox: `yes`
- version: `nix-env (Nix) 2.3.4`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: 
- rustplatform.buildRustPackage
```

**Comment by Mic92** [comment]
Related to https://github.com/NixOS/nixpkgs/pull/82496

**Comment by makefu** [comment]
still an issue for me!

**Comment by makefu** [comment]
still an issue for me

**Comment by Mic92** [comment]
@makefu I think you found the solution in the end, right?

**Comment by makefu** [comment]
@Mic92 right, i somehow found a solution for the proxying issue, however the intercepting proxy is also messing with the source files when building from source. For reference, this is the solution i came up with:

```nix
let
  fetchCargoTarball = rustPlatform.fetchCargoTarball.override { cargo = cargo.override { cacert = my-cacert;}; };
in
(rustPlatform.buildRustPackage.override { inherit fetchCargoTarball;  }) rec {
...
```
this will result in a rebuild of the rustPlatform with the injected cacert file. However it would be great if this would not be necessary in first place and cargo would honor the ca-bundle set via `SSL_CERT_FILE`

**Comment by linsui** [comment]
A easier workaround which doesn't rebuild the cargo:

```nix
  buildRustPackage = rustPlatform.buildRustPackage.override {
    fetchCargoTarball = rustPlatform.fetchCargoTarball.override {
      cargo = writeShellScriptBin "cargo" ''
        export CARGO_HTTP_CAINFO="${cacert}/etc/ssl/certs/ca-bundle.crt"
        export SSL_CERT_FILE="${cacert}/etc/ssl/certs/ca-bundle.crt"
        ${cargo}/bin/.cargo-wrapped "$@"
      '';
    };
  };
```

**Comment by zowoq** [comment]
Should be resolved by https://github.com/NixOS/nixpkgs/pull/210366

**Comment by malik-n** [comment]
> Should be resolved by [#210366](https://github.com/NixOS/nixpkgs/pull/210366)

Still experiencing this issue
